File: RB101_Programming_Foundations/small_problems/Easy4/what_century.rb
# Input: year
# output century string that ends with st nd rd th as appropriate for that number

#logic
  # To get the current century number
  #   Base is the current year divided by 100, integer division
  #   Check if the number ends with 00, convert to string and compare the string values
  #     if true, don't increment the base by 1, else increment the base by 1

  # Append suffix
  #     if the second to last century digit is 1, append everything to 'th'
  #                         else
  #                           If the last century digit is
  #                                  1, append "st
  #                                  2, append end
  #                                  3. append rd
  #                             else append 'th'


def century(year)
  base_century = year / 100
  remain_years = year % 100
  century = remain_years == 0 ? base_century : base_century + 1

  century.to_s + century_suffix(century)

end

def century_suffix(century)

  if !([11, 12, 13].include?(century % 100))
    last_digit = century % 10

    case last_digit
    when 1
      'st'
    when 2
      'nd'
    when 3
      'rd'
    else
     'th'
    end
  else
    'th'
  end
end







puts century(2000) == '20th'
puts century(2001) == '21st'
puts century(1965) == '20th'
puts century(256) == '3rd'
puts century(5) == '1st'
puts century(10103) == '102nd'
puts century(1052) == '11th'
puts century(1127) == '12th'
puts century(11201) == '113th'

File: js210_fundamentals_of_JS_Programming/problem_sets/easy2/what_century.js
/*
Input: a year 
Output: the century the year is in 
Rules: 
  if the century number is appended with th unless
    the century ends in 1, 2, 3, which ends in st, nd, rd, respectively, unless
    the last two digits of the century is 11, 12, and 13



Data Structure:
  

Algorithms:
  convert the year into century by integer division by 100
  if there is remainder, then increment the century number by 1

  if the century's last two digits are 11, 12, and 13, append th
  if the century's last digit is 1, append st, 2, nd, 3 rd
  else append th
  
  returen the result
*/

function century(year) {
  let centuryNum = Math.floor(year / 100);
  if (year % 100 !== 0) centuryNum += 1;

  return String(centuryNum) + centurySuffix(centuryNum);
}

function centurySuffix(centuryNumber) {
  let lastTwoDigits = centuryNumber % 100;
  let lastDigit = centuryNumber % 10;

  if ([11, 12, 13].indexOf(lastTwoDigits) !== -1) return 'th';

  switch (lastDigit) {
    case 1:
      return 'st';
    case 2:
      return 'nd';
    case 3:
      return 'rd';
    default:
      return 'th';
  }
}

console.log(century(2000)); // "20th"
console.log(century(2001)); // "21st"
console.log(century(1965)); // "20th"
console.log(century(256)); // "3rd"
console.log(century(5)); // "1st"
console.log(century(10103)); // "102nd"
console.log(century(1052)); // "11th"
console.log(century(1127)); // "12th"
console.log(century(11201)); // "113th"
